pkg/querier/queryrange/downstreamer.go

```diff
@@ -249,3 +258,374 @@ func ResponseToResult(resp queryrangebase.Response) (logqlmodel.Result, error) {
 	}
 }
+
+// downstreamAccumulator is one of two variants:
+// a logsAccumulator or a bufferedAccumulator.
+// Which variant is detected on the first call to Accumulate.
+// Metric queries, which are generally small payloads, are buffered
+// since the memory overhead is negligible.
+// Log queries, sharded thousands of times and each returning <limit>
+// results, can be _considerably_ larger.	In this case, we eagerly
+// accumulate the results into a logsAccumulator, discarding values
+// over the limit to keep memory pressure down while other subqueries
+// are executing.
+type downstreamAccumulator struct {
+	acc    resultAccumulator
+	params logql.Params
+	n      int // number of queries, used to build slice size
+}
+
+type resultAccumulator interface {
+	Accumulate(logqlmodel.Result, int) error
+	Result() []logqlmodel.Result
+}
+
+func newDownstreamAccumulator(params logql.Params, nQueries int) *downstreamAccumulator {
+	return &downstreamAccumulator{params: params, n: nQueries}
+}
+
+func (a *downstreamAccumulator) build(acc logqlmodel.Result) {
+	if acc.Data.Type() == logqlmodel.ValueTypeStreams {
+
+		// the stream accumulator stores a heap with reversed order
+		// from the results we expect, so we need to reverse the direction
+		direction := logproto.FORWARD
+		if a.params.Direction() == logproto.FORWARD {
+			direction = logproto.BACKWARD
+		}
+
+		a.acc = newStreamAccumulator(direction, int(a.params.Limit()))
+
+	} else {
+		a.acc = &bufferedAccumulator{
+			results: make([]logqlmodel.Result, a.n),
+		}
+
+	}
+}
+
+func (a *downstreamAccumulator) Accumulate(ctx context.Context, index int, acc logqlmodel.Result) error {
+	// on first pass, determine which accumulator to use
+	if a.acc == nil {
+		a.build(acc)
+	}
+
+	return a.acc.Accumulate(acc, index)
+}
+
+func (a *downstreamAccumulator) Result() []logqlmodel.Result {
+	if a.acc == nil {
+		return nil
+	}
+	return a.acc.Result()
+
+}
+
+type bufferedAccumulator struct {
+	results []logqlmodel.Result
+}
+
+func (a *bufferedAccumulator) Accumulate(acc logqlmodel.Result, i int) error {
+	a.results[i] = acc
+	return nil
+}
+
+func (a *bufferedAccumulator) Result() []logqlmodel.Result {
+	return a.results
+}
+
+// heap impl for keeping only the top n results across m streams
+// importantly, accumulatedStreams is _bounded_, so it will only
+// store the top `limit` results across all streams.
+// To implement this, we use a min-heap when looking
+// for the max values (logproto.FORWARD)
+// and vice versa for logproto.BACKWARD.
+// This allows us to easily find the 'worst' value
+// and replace it with a better one.
+// Once we've fully processed all log lines,
+// we return the heap in opposite order and then reverse it
+// to get the correct order.
+// Heap implements container/heap.Interface
+// solely to use heap.Interface as a library.
+// It is not intended for the heap pkg functions
+// to otherwise call this type.
+type accumulatedStreams struct {
```

Comment:
How do you feel about using Bryan's Loser Tree implementation https://github.com/grafana/loki/blob/main/pkg/util/loser/tree.go?